torch/csrc/jit/autodiff.cpp

```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+      if (node->stage() != stage) break;
+      for (Value * output : node->outputs())
+        order[output] = next_nr++;
+    }
+    // NOTE: outputs are either inputs or are produced by nodes,
+    // so we don't need to handle them in any way.
+    stage++;
+  }
+  JIT_ASSERT(stage == graph->stage() + 1);
+  return order;
+}
+
+LiftedReverse lambdaLiftReverse(std::shared_ptr<Graph>& graph) {
+  static const auto is_stage_0 = [](Value *v) { return v->stage() == 0; };
+  static const auto is_stage_1 = [](Value *v) { return v->stage() == 1; };
+  // NOTE: in the comments here first stage == stage 0
+  JIT_ASSERT(graph->stage() == 1);
+
+  // --------------------------------------------------------------------------
+  // 1. Find values of stage 0 that need to ba captured.
+  // --------------------------------------------------------------------------
+  // First, we need to find all values that are produced in the first stage,
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  std::unordered_set<Value*> extra_reverse_inputs;
+  for (Node * node : *graph) {
+    if (node->stage() == 0) continue;
+    for (Value * input : node->inputs()) {
+      if (input->stage() != 0) continue;
+      extra_reverse_inputs.insert(input);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // 2. Establish an ordering on inputs/outputs.
+  // --------------------------------------------------------------------------
+  // We will need to sort primal outputs and reverse inputs, so that they maintain
+  // an order of:
+  //   [primal outputs], [primal temporaries], [reverse inputs]
+  // where each section is sorted in the same way as in the input graph.
+  //
+  // This corresponds exactly to taking a topological ordering on values,
+  // but artificially boosting outputs of the first stage to precede all
+  // other values.
+  auto topo_order = computeValueTopoOrder(graph);
+  auto io_compare = [&](Value *a, Value *b) { return topo_order.at(a) < topo_order.at(b); };
+  { // Boost outputs of primal graph
+    int64_t next_nr = -static_cast<int64_t>(graph->outputs().size());
+    for (Value * output : graph->outputs()) {
+      if (output->stage() != 0) break;
+      topo_order[output] = next_nr++;
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // 3. Prepare input/outputs lists for both graphs.
+  // --------------------------------------------------------------------------
+  // It's simple to construct primal_inputs/reverse_outputs,
+  // but primal_outputs/reverse_inputs are much more subtle.
+  // Here's a summary of how they are supposed to look like:
+  //
+  // Primal outputs:
+  //   [original outputs], [temporaries]
+  //
+  // Reverse inputs:
+  //   [original outputs#], [temporaries], [output vjps (aka grad_outputs)]
```

Comment:
We need `grad_outputs` for the `[temporaries]` list as well, and the gradient calculation needs to be updated to add those values into the grad_temp variable in the reverse. This likely means that you will need to still have the `grad_map` around when you do this, and that this code need to be reorganized to make that possible.

Reply:
Good point. Will need to update that 😕